### ISSUE TITLE
Exporting encrypted vault with Personal API Key?

### ISSUE THREAD

**Issue by berkant** [bitwarden/clients#2739]
Hello.

Is there a way to export an encrypted vault after logging in with `--apikey` flag.

The scenario is running a periodic task with GitHub Actions to export and re-upload my encrypted vault to another storage medium. I thought I could call `export` without a master password but the case seems not. I don't feel like providing GitHub with my master password.

I just noticed `bw` populates `.config/Bitwarden\ CLI/data.json` with encrypted stuff seemingly including my logins, but I am not sure if I can restore using that file sometime in the future. Is data exported by `bw export` any different than `data.json`?

---------

Opinion: You should probably back up `data.json` instead of encrypted exports. Encrypted exports don't include your encrypted private key and other crucial data used to decrypt your vault. That means you will have to connect to Bitwarden's servers to retrieve them, which is obviously not ideal to many.

**Comment by TomiBelan** [comment]
I looked into this issue and I think I understand it pretty well now. But I'm just a user so I welcome corrections from the Bitwarden team.

**Q: Why does `bw export --format=encrypted_json` even need the master password?**

A: In old CLI versions, bw export didn't truly need it. It asked for the password but didn't actually use it for anything. I have an old patched build of bw 1.15.1 which just deletes the password prompt code from export.command.ts and vault.program.ts, and it used to work just fine.

Nowadays it's not so simple. If I try to delete the relevant exitIfLocked() call, bw export crashes with `Cannot read properties of null (reading 'encKey')`. This is because of bitwarden/jslib#369, which added a new field named `encKeyValidation_DO_NOT_EDIT` to the encrypted_json format. Everything else in the encrypted_json output could be produced without having the master password, only this single field is an obstacle.

The field contains a random GUID encrypted with the user's account encryption key. It is used as an integrity check during import, to verify that the account encryption key still matches. Without this check, importing an encrypted_json file to a different account (or same account after rotating the account encryption key) could lead to silent corruption. See also [Encrypted Exports docs](https://bitwarden.com/help/encrypted-export/).

**Q: How to change `bw export` to allow using it without the master password?**

A: One possible idea could be to add a new option `--without-integrity-check` which omits the `encKeyValidation_DO_NOT_EDIT` field from the output. Such json files are valid because the import command simply skips the key validation if the field is missing. But it increases the risk of issues from account encryption key rotation, so this mode shouldn't be the default for normal users when they make encrypted_json backups manually.

Then the code could be modified so that if `--without-integrity-check` is present, and format equals `encrypted_json`, and the `--password` option was not given, `bw export` won't ask for the master password and won't unlock the vault.

I could write a PR if there is interest, but I don't like this idea very much.

**Q: Is it even desirable to generate encrypted_json exports automatically and use them as backups?**

A: I'm not so sure. Even if you could run `bw export` without a master password prompt, the output is still tied to your current account encryption key. Per [Encrypted Exports docs](https://bitwarden.com/help/encrypted-export/), the file can't be imported into a different account, and it'll stop working if you ever rotate the account encryption key.

By contrast, the data.json file is self-contained. Even if all your devices and all Bitwarden servers explode in the same instant, as long as you know your master password and you have a backup of data.json, you'll be able to get your passwords back. And although technically data.json is "not guaranteed to be 1:1 to an export" (e.g. you can't `bw import` it directly into another Bitwarden account), you can run `BITWARDENCLI_APPDATA_DIR=/path/to/backup bw export --type=json` manually during restore.

Of course, your choice may depend on your threat model and whether this is something you want.

(Explanation: The account encryption key is a random number initially generated when you register. Bitwarden client encrypts it with your master password and saves the encrypted form in data.json and on Bitwarden servers. It is not saved in encrypted_json. When you rotate it, Bitwarden generates a new random number, re-encrypts all saved passwords with it, and saves the new number encrypted with your new master password in data.json and on Bitwarden servers.)

**Q: What about attachments?**

A: Sorry, I don't know anything about them.

**Comment by Alexander-Shukaev** [comment]
@TomiBelan, I guess in this case, in order for encrypted export feature to remain reasonably portable, it's unavoidable to also export the encryption key (in the encrypted from of course) as a separate file or maybe as part of encrypted JSON itself.  So the only protection left would be the master password, which should be strong anyway regardless of this feature.  Additionally, it probably makes sense to have a separate CLI sub-command that instead of directly importing encrypted JSON, merely decrypts it and writes decrypted JSON with all the vault data to standard output.  This sounds portable enough to me, reducing vendor lock on Bitwarden infrastructure, except for encryption/decryption algorithm.

**Comment by TomiBelan** [comment]
Right, that matches my understanding. Personally, my conclusion is the same as berkant's. I just make backups of the data.json file. I want a format that can be decrypted completely offline using only my master password, even if Bitwarden servers are down or if I lose access to my account, and data.json is exactly that.

In my case I'm using `bw sync` instead of `bw login --apikey`, but that's up to personal preference. Either way it works without the master password or system keyring. `bw sync` doesn't need the password since bw 2022.9.0 as long as you're logged in.

<details><summary>My backup script</summary>

Keep in mind that this script does not handle Bitwarden attachments!

```bash
#!/bin/bash
set -e
IFS=$'\n'
here=$(dirname "$0")
export BITWARDENCLI_APPDATA_DIR="$here/appdata"
today=$(date +%Y-%m-%d)

# Keep last 50 backups
rm -rf $(printf "%s\n" "$here"/backups/* | sort | head -n-50)

if ! [[ -e "$here/backups/$today" ]]; then
  # This is just a hack to get rid of the "Syncing complete." success message, but preserve any errors.
  if out=$(bw sync 2>&1); then
    true
  else
    res=$?
    echo "bw sync: failed with $res" >&2
    echo "$out" >&2
    exit $res
  fi

  mkdir "$here/backups/$today"
  cp -a "$here/appdata/"* "$here/backups/$today/"
fi
```

</details>

For data.json-based backups, you could say "bw export" already fulfills the role of "separate CLI sub-command that instead of directly importing encrypted JSON, merely decrypts it". That's another advantage over format=encrypted_json.

**Comment by closebot-bw** [comment]
⚠️ **Stale Issue Notice**

This issue has been automatically marked as stale due to inactivity. It will be closed in **2 weeks** (October 15, 2025) if no further activity occurs.

If this issue is still relevant and you would like to keep it open, please:
- Comment on this issue to show continued interest
- Provide any additional information or updates
- Confirm that the issue still exists in the latest version

Thank you for your contribution to this project! 🙏

**Comment by closebot-bw** [comment]
🔒 **Issue Closed Due to Inactivity**

This issue has been automatically closed due to lack of activity for an extended period. We periodically review and close inactive issues to help maintain our issue tracker and focus on current priorities.

**If this issue is still relevant:**
- Please create a new issue with updated information
- Include steps to reproduce the problem if it's a bug report
- Mention if this issue still occurs in the latest version

Thank you for your contribution to this project. Your feedback helps us improve! 🙏
